host/xtest/adbg/src/adbg_run.c

```diff
@@ -130,17 +130,28 @@ static int ADBG_RunSuite(
 
 	TAILQ_FOREACH(case_def, &Runner_p->Suite_p->cases, link) {
 		if (argc > 0) {
-			bool HaveMatch = false;
+			bool check_inclusion = false;
+			bool match_inclusion = false;
+			bool match_exclusion = false;
 			int i = 0;
 
 			for (i = 0; i < argc; i++) {
-
-				if (strstr(case_def->TestID_p, argv[i])) {
-					HaveMatch = true;
-					break;
+				if (!strcmp(argv[i], "-x")) {
+					i++;
+					if (i >= argc) {
+						Do_ADBG_Log("Error: -x ARG");
+						return 1;
+					}
+					if (strstr(case_def->TestID_p, argv[i]))
+						match_exclusion = true;
+				} else {
+					check_inclusion = true;
+					if (strstr(case_def->TestID_p, argv[i]))
+						match_inclusion = true;
 				}
 			}
-			if (!HaveMatch) {
+			if (match_exclusion ||
+			    (check_inclusion && !match_inclusion)) {
```

Comment:
You don't need several booleans. How about this instead?
```diff
diff --git a/host/xtest/adbg/src/adbg_run.c b/host/xtest/adbg/src/adbg_run.c
index 8d4e31f..17adbbf 100644
--- a/host/xtest/adbg/src/adbg_run.c
+++ b/host/xtest/adbg/src/adbg_run.c
@@ -129,21 +129,29 @@ static int ADBG_RunSuite(
        Do_ADBG_Log("######################################################");

        TAILQ_FOREACH(case_def, &Runner_p->Suite_p->cases, link) {
+               bool HaveMatch = false;
+
                if (argc > 0) {
-                       bool HaveMatch = false;
                        int i = 0;

                        for (i = 0; i < argc; i++) {
-
-                               if (strstr(case_def->TestID_p, argv[i])) {
-                                       HaveMatch = true;
-                                       break;
+                               if (!strcmp(argv[i], "-x")) {
+                                       i++;
+                                       if (i >= argc) {
+                                               Do_ADBG_Log("Error: -x ARG");
+                                               return 1;
+                                       }
+                                       if (strstr(case_def->TestID_p, argv[i]))
+                                               HaveMatch = false;
+                               } else {
+                                       if (strstr(case_def->TestID_p, argv[i]))
+                                               HaveMatch = true;
                                }
                        }
-                       if (!HaveMatch) {
-                               NumSkippedTestCases++;
-                               continue;
-                       }
+               }
+               if (!HaveMatch) {
+                       NumSkippedTestCases++;
+                       continue;
                }

                Case_p = ADBG_Case_New(case_def);
```

WIth the advantage that order becomes significant which gives more possibilities to include and exclude tests.

Reply:
I understand your sequence but it fails when only `-x ID` are provided or when no ID is passed.
I'll see how to fix that.

Reply:
Ah yes. That should fix the problematic cases:
```diff
diff --git a/host/xtest/adbg/src/adbg_run.c b/host/xtest/adbg/src/adbg_run.c
index 17adbbf..9ef99ce 100644
--- a/host/xtest/adbg/src/adbg_run.c
+++ b/host/xtest/adbg/src/adbg_run.c
@@ -129,7 +129,7 @@ static int ADBG_RunSuite(
 	Do_ADBG_Log("######################################################");
 
 	TAILQ_FOREACH(case_def, &Runner_p->Suite_p->cases, link) {
-		bool HaveMatch = false;
+		bool HaveMatch = !argc || !strcmp(argv[0], "-x");
 
 		if (argc > 0) {
 			int i = 0;
```

Reply:
I just saw your latest proposal...  P-R updated accordingly.